### PR TITLE
[CORDA-2394]: Fixed ABI issue with VaultQueryCriteria.

### PR DESCRIPTION
**Pull request by sollecitom**
https://r3-cev.atlassian.net/browse/CORDA-2394

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt [hunk 1]
```diff
@@ -94,12 +94,39 @@ sealed class QueryCriteria : GenericQueryCriteria<QueryCriteria, IQueryCriteriaP
...
+                constraints: Set<Vault.ConstraintInfo> = emptySet(),
+                participants: List<AbstractParty>? = null
+        ) : this(status, contractStateTypes, stateRefs, notary, softLockingCondition, timeCondition) {
+            this.relevancyStatus = relevancyStatus
+            this.constraintTypes = constraintTypes
+            this.constraints = constraints
+            this.participants = participants
+        }
+
+        override var relevancyStatus: Vault.RelevancyStatus = Vault.RelevancyStatus.ALL
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt]: Why not leave the primary c'tor with all the properties and then have this additional V3 c'tor (with the same default values)? That should avoid the need to have these private setters.

**Comment by Alex-Koller-R3** [comment]
The fix has resolved the problem I was seeing with BNMS

File: core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt (hunk 1 above)
Comment (sollecitom) [reply core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt]: As discussed f2f, that doesn't work.

File: core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt [hunk 2]
```diff
@@ -249,12 +296,14 @@ sealed class QueryCriteria : GenericQueryCriteria<QueryCriteria, IQueryCriteriaP
         fun copy(
                 expression: CriteriaExpression<L, Boolean> = this.expression,
                 status: Vault.StateStatus = this.status,
-                contractStateTypes: Set<Class<out ContractState>>? = this.contractStateTypes
+                contractStateTypes: Set<Class<out ContractState>>? = this.contractStateTypes,
+                relevancyStatus: Vault.RelevancyStatus = this.relevancyStatus
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt]: Why is this added?

Comment (sollecitom) [reply core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt]: Because it was removed by the primary constructor of the data class, so the "primary" copy is now the one without, and this is putting the one with additional fields back in place.

Comment (shamsasari) [reply core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt]: Yes, sorry I was confused by the other class.
